R/netzschleuder.R

```diff
@@ -0,0 +1,205 @@
+        "i" = "see {.url {collection_url}}"
+      )
+    )
+  } else if (net_ident[1] == net_ident[2]) {
+    return(raw)
+  } else {
+    idx <- which(unlist(raw[["nets"]]) == net_ident[2])
+    if (length(idx) == 0) {
+      cli::cli_abort(
+        c(
+          "{net_ident[2]} is not part of the collection {net_ident[1]}.",
+          "i" = "see {.url {collection_url}}"
+        )
+      )
+    }
+    raw[["analyses"]] <- raw[["analyses"]][[net_ident[2]]]
+    raw[["nets"]] <- raw[["nets"]][idx]
+    raw
+  }
+}
+
+#' @rdname netzschleuder
+#' @export
+ns_df <- function(name, token = NULL) {
+  if (!requireNamespace("minty", quietly = TRUE)) {
+    cli::cli_abort(
+      "The package `minty` is needed for this function. Please install it."
+    )
+  }
+  meta <- ns_metadata(name)
+  net_ident <- resolve_name(name)
+
+  zip_url <- sprintf(
+    "net/%s/files/%s.csv.zip",
+    net_ident[1],
+    net_ident[2]
+  )
+
+  temp <- tempfile(fileext = "zip")
+  download_file(zip_url, token = token, file = temp)
+
+  zip_contents <- utils::unzip(temp, list = TRUE)
+
+  edge_file_name <- grep("edge", zip_contents$Name, value = TRUE)
+  node_file_name <- grep("node", zip_contents$Name, value = TRUE)
+  gprops_file_name <- grep("gprops", zip_contents$Name, value = TRUE)
+
+  edges_df_raw <- utils::read.csv(unz(temp, edge_file_name))
+  edges_df <- suppressWarnings(minty::type_convert(edges_df_raw))
+  source_loc <- grep("source", names(edges_df))
+  target_loc <- grep("target", names(edges_df))
+
+  # netzschleuder uses 0-indexing, igraph uses 1-indexing
+  edges_df[[source_loc]] <- edges_df[[source_loc]] + 1
+  edges_df[[target_loc]] <- edges_df[[target_loc]] + 1
+  names(edges_df)[c(source_loc, target_loc)] <- c("from", "to")
+
+  nodes_df_raw <- utils::read.csv(unz(temp, node_file_name))
+  #suppress warning if no character columns found
+  nodes_df <- suppressWarnings(minty::type_convert(nodes_df_raw))
+  names(nodes_df)[1] <- "id"
+
+  # netzschleuder uses 0-indexing, igraph uses 1-indexing
+  nodes_df[["id"]] <- nodes_df[["id"]] + 1
+  if ("X_pos" %in% names(nodes_df)) {
+    pos_array <- gsub("array\\(\\[|\\]|\\)", "", nodes_df[["X_pos"]])
+    split_coords <- strsplit(pos_array, ",")
+
+    x_vals <- vapply(
+      split_coords,
+      function(x) as.numeric(trimws(x[1])),
+      numeric(1)
+    )
+    y_vals <- vapply(
+      split_coords,
+      function(x) as.numeric(trimws(x[2])),
+      numeric(1)
+    )
+
+    nodes_df[["X_pos"]] <- NULL
+    nodes_df[["x"]] <- x_vals
+    nodes_df[["y"]] <- y_vals
+  }
+
+  #gprops file is not always a valid csv file. In that case, simply read it as text
+  gprops_df <- readLines(unz(temp, gprops_file_name))
+
+  on.exit(unlink(temp))
+
+  list(nodes = nodes_df, edges = edges_df, gprops = gprops_df, meta = meta)
+}
+
+#' @rdname netzschleuder
+#' @export
+ns_graph <- function(name, token = NULL) {
+  graph_data <- ns_df(name, token = token)
```

Comment:
My comment was not about S3, we can implement poor man's dispatch here (hand-rolled check if it's a character or a classed object). Users may want to inspect or do other things to the meta object before creating a graph or a data frame from it. We could also decouple downloading from df creation, but that's a bit of a can of worms.